File: iRHSCBook/RHSCFindMemberViewController.m
//
//  RHSCFindMemberViewControllerTableViewController.m
//  iRHSCBook
//
//  Created by Bruce Hunter on 2014-06-30.
//  Copyright (c) 2014 Richmond Hill Squash Club. All rights reserved.
//

#import "RHSCFindMemberViewController.h"
#import "RHSCTabBarController.h"

@interface RHSCFindMemberViewController () <UISearchDisplayDelegate,UISearchBarDelegate>

@property (nonatomic,strong) NSMutableArray *filteredList;
@property (nonatomic,weak) IBOutlet UITableView *searchResultsView;
@property (nonatomic,weak) IBOutlet UITableView *memberListView;

@end

@implementation RHSCFindMemberViewController

BOOL searching;

- (id)initWithStyle:(UITableViewStyle)style
{
    self = [super initWithStyle:style];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
    self.searchDisplayController.displaysSearchBarInNavigationBar = YES;
    RHSCTabBarController *tbc = (RHSCTabBarController *)self.tabBarController;
    RHSCMemberList *ml = tbc.memberList;
    self.filteredList = [[NSMutableArray alloc] initWithArray:ml.memberList];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    if(self.searchDisplayController.searchResultsTableView == tableView) {
        return self.filteredList.count;
    } else {
        // get memberList count
        RHSCTabBarController *tbc = (RHSCTabBarController *)self.tabBarController;
        RHSCMemberList *ml = tbc.memberList;
        return ml.memberList.count;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    
    if(self.searchDisplayController.searchResultsTableView == tableView) {
        UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"MemberCell" forIndexPath:indexPath];
        RHSCMember *member = self.filteredList[indexPath.row];
        cell.textLabel.text = [NSString stringWithFormat:@"%@, %@",member.lastName,member.firstName];
        return cell;
    } else {
        UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"MemberCell" forIndexPath:indexPath];
    
        // Configure the cell...
        RHSCTabBarController *tbc = (RHSCTabBarController *)self.tabBarController;
        RHSCMemberList *ml = tbc.memberList;
        RHSCMember *member = ml.memberList[indexPath.row];
        cell.textLabel.text = [NSString stringWithFormat:@"%@, %@",member.lastName,member.firstName];
        return cell;
    
    }
}

- (void) searchTableView
{
    NSString *searchText = self.searchDisplayController.searchBar.text;
    
    RHSCTabBarController *tbc = (RHSCTabBarController *)self.tabBarController;
    RHSCMemberList *ml = tbc.memberList;
    for (RHSCMember *item in ml.memberList) {
        NSRange range = [item.name rangeOfString:searchText options:NSCaseInsensitiveSearch];
        if (range.length > 0) {
            [self.filteredList addObject:item];
        }
    }
}

- (void)searchDisplayControllerWillBeginSearch:(UISearchDisplayController *)controller
{
    searching = NO;
}

- (void)searchDisplayControllerWillEndSearch:(UISearchDisplayController *)controller
{
    searching = NO;
    [self.tableView reloadSections:[NSIndexSet indexSetWithIndex:0]
                  withRowAnimation:UITableViewRowAnimationAutomatic];
    [self.filteredList removeAllObjects];
}

- (BOOL)searchDisplayController:(UISearchDisplayController *)controller
shouldReloadTableForSearchString:(NSString *)searchText
{
    [self.filteredList removeAllObjects];
    
    if ([searchText length] > 1) {
        searching = YES;
        [self searchTableView];
    } else {
        searching = NO;
    }
    return YES;
}


/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath
{
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: iRHSCBook/RHSCFindMemberViewController.swift
//
//  RHSCFindMemberViewController.swift
//  iRHSCBook
//
//  Created by Bruce Hunter on 2015-10-30.
//  Copyright © 2015 Bruce Hunter. All rights reserved.
//

import Foundation
import UIKit
fileprivate func < <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
  switch (lhs, rhs) {
  case let (l?, r?):
    return l < r
  case (nil, _?):
    return true
  default:
    return false
  }
}

fileprivate func > <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
  switch (lhs, rhs) {
  case let (l?, r?):
    return l > r
  default:
    return rhs < lhs
  }
}


protocol setPlayerProtocol {

    func setPlayer(_ player: RHSCMember?, playerNumber: UInt16)

}

class RHSCFindMemberViewController : UITableViewController,UISearchResultsUpdating {

    var playerNumber : UInt16 = 0
    var delegate : AnyObject? = nil
    
    var filteredList : Array<RHSCMember> = []
    var resultSearchController : UISearchController!
    
    @IBOutlet weak var searchResultsView : UITableView? = nil
    @IBOutlet weak var memberListView : UITableView? = nil
    
    var searching : Bool = false

    override func viewDidLoad() {
        super.viewDidLoad()
    
        self.tableView.accessibilityIdentifier = "FindMember"
        // Uncomment the following line to preserve selection between presentations.
        // self.clearsSelectionOnViewWillAppear = NO;
    
        // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
        // self.navigationItem.rightBarButtonItem = self.editButtonItem;
        //self.searchDisplayController.displaysSearchBarInNavigationBar = YES;
        let tbc = self.tabBarController as! RHSCTabBarController
        let ml = tbc.memberList
        
        self.filteredList = []
        self.filteredList.append(contentsOf: ml!.memberList)
        
        resultSearchController = UISearchController(searchResultsController: nil)
        resultSearchController.searchResultsUpdater = self
        resultSearchController.hidesNavigationBarDuringPresentation = false
        resultSearchController.dimsBackgroundDuringPresentation = false
        resultSearchController.searchBar.searchBarStyle = UISearchBarStyle.prominent
        resultSearchController.searchBar.sizeToFit()
        self.tableView.tableHeaderView = resultSearchController.searchBar
        
        self.tableView.reloadData()
    }
    
    func updateSearchResults(for searchController: UISearchController) {
        if searchController.searchBar.text?.characters.count > 0 {
            filteredList.removeAll(keepingCapacity: false)
            let searchPredicate = NSPredicate(format: "self.fullName contains[c] %@", searchController.searchBar.text!)
            let tbc = self.tabBarController as! RHSCTabBarController
            let array = (tbc.memberList!.memberList as NSArray).filtered(using: searchPredicate)
            filteredList = array as! [RHSCMember]
            tableView.reloadData()
        }
        else {
            filteredList.removeAll(keepingCapacity: false)
            let tbc = self.tabBarController as! RHSCTabBarController
            filteredList = tbc.memberList!.memberList
            tableView.reloadData()
        }
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        if (resultSearchController.isActive)
        {
            return self.filteredList.count
        }
        else
        {
            let tbc = self.tabBarController as! RHSCTabBarController
            return tbc.memberList!.memberList.count
        }
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
    {
        let cell = tableView.dequeueReusableCell(withIdentifier: "MemberCell", for: indexPath) as UITableViewCell
        
        if (resultSearchController.isActive)
        {
            let mem = self.filteredList[(indexPath as NSIndexPath).row]
            cell.textLabel?.text = mem.sortName
        }
        else
        {
            let tbc = self.tabBarController as! RHSCTabBarController
            let mem = tbc.memberList!.memberList[(indexPath as NSIndexPath).row]
            cell.textLabel?.text = mem.sortName
        }
        cell.textLabel!.backgroundColor = UIColor.clear
        cell.textLabel!.textColor = UIColor.black
        cell.accessoryType = .none
        cell.accessibilityIdentifier = cell.textLabel?.text
        return cell
    }

    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        cell.contentView.backgroundColor = UIColor.red
    }
    
   
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        //    NSLog(@"popping FindMember on table select");
        let selectedIndexPath = tableView.indexPathForSelectedRow
        var selmem : RHSCMember? = nil
        if (resultSearchController.isActive) {
            selmem = self.filteredList[(selectedIndexPath! as NSIndexPath).row]
        } else {
            let tbc = self.tabBarController as! RHSCTabBarController
            let ml = tbc.memberList
            selmem = ml!.memberList[(selectedIndexPath! as NSIndexPath).row]
        }
        if delegate is RHSCBookCourtViewController {
            let deltarget = (delegate as! RHSCBookCourtViewController)
            deltarget.setPlayer(selmem, number: playerNumber)
        }
        if delegate is RHSCUpdateCourtViewController {
            let deltarget = (delegate as! RHSCUpdateCourtViewController)
            deltarget.setPlayer(selmem, number: playerNumber)
        }
        resultSearchController.isActive = false
        self.navigationController?.popViewController(animated: false)
    }

}
